src/cargo/core/compiler/build_runner/mod.rs

```diff
@@ -302,6 +303,49 @@ impl<'a, 'gctx> BuildRunner<'a, 'gctx> {
                     .insert(dir.clone().into_path_buf());
             }
         }
+
+        if self.bcx.build_config.intent.is_doc()
+            && self.bcx.gctx.cli_unstable().rustdoc_mergeable_info
+            && let Some(unit) = self
+                .bcx
+                .roots
+                .iter()
+                .filter(|unit| unit.mode.is_doc())
+                .next()
```

Comment:
Is the finial merge step required to be run always?

I was wondering the fingerprint/caching story from here <https://github.com/rust-lang/cargo/issues/16291>. Perhaps a better model is making this a `Unit` with dependencies of all rustdoc invocations.

(Happy to help continue Cargo integration if you need more time on rustdoc side 🙂)

Reply:
> Is the finial merge step required to be run always?

The final merge step has to run if any of the docs have been changed. It can be skipped only if absolutely none of the crates needed documented.

> Perhaps a better model is making this a Unit with dependencies of all rustdoc invocations.

I would love to. Unfortunately, the merge step is not associated with any crate in particular, and `Unit` has a non`Option`al [package](https://github.com/rust-lang/cargo/blob/b0896a0c330009034ac8b403c4fa35c23e167337/src/cargo/core/compiler/unit.rs#L43) member variable.

> (Happy to help continue Cargo integration if you need more time on rustdoc side 🙂)

Thank you!